core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -300,106 +308,55 @@ protected BlockInStream createUnderStoreBlockInStream(long blockStart, long leng
-   * {@link #mShouldCacheCurrentBlock} is set to {@link #mAlluxioStorageType}.isCache().
-   * {@link #mCurrentCacheStream} is also closed and a new one is created for the next block.
+   * Check whether block instream and cache outstream should be updated.
+   * This function is only called by {@link #updateStreams()}.
    *
-   * @throws IOException if the next BlockInStream cannot be obtained
+   * @param currentBlockId cached result of {@link #getCurrentBlockId()}
+   * @return true if the block stream should be updated
    */
-  private void checkAndAdvanceBlockInStream() throws IOException {
-    long currentBlockId = getCurrentBlockId();
-    if (mCurrentBlockInStream == null
-        || (mCurrentBlockInStream.remaining() == 0 && currentBlockId != mCurrentBlockId)) {
-      closeCacheStream();
-      updateBlockInStream(currentBlockId);
-      mCurrentBlockId = currentBlockId;
-      if (mShouldCacheCurrentBlock) {
-        try {
-          WorkerNetAddress address = mLocationPolicy.getWorkerForNextBlock(
-              mContext.getAlluxioBlockStore().getWorkerInfoList(), getBlockSizeAllocation(mPos));
-          // Don't cache the block to somewhere that already has it.
-          // TODO(andrew): Filter the workers provided to the location policy to not include
-          // workers which already contain the block. See ALLUXIO-1816.
-          if (mCurrentBlockInStream instanceof RemoteBlockInStream) {
-            WorkerNetAddress readAddress =
-                ((RemoteBlockInStream) mCurrentBlockInStream).getWorkerNetAddress();
-            // Try to avoid an RPC.
-            if (readAddress.equals(address)) {
-              mShouldCacheCurrentBlock = false;
-            } else {
-              BlockInfo blockInfo = mContext.getAlluxioBlockStore().getInfo(currentBlockId);
-              for (BlockLocation location : blockInfo.getLocations()) {
-                if (address.equals(location.getWorkerAddress())) {
-                  mShouldCacheCurrentBlock = false;
-                }
-              }
-            }
-          }
-          if (mShouldCacheCurrentBlock) {
-            mCurrentCacheStream = createCacheStream(currentBlockId, getBlockSize(mPos), address);
-          }
-        } catch (IOException e) {
-          logCacheStreamIOException(e);
-          mShouldCacheCurrentBlock = false;
-        } catch (AlluxioException e) {
-          LOG.warn(BLOCK_ID_NOT_CACHED, currentBlockId, e);
-          mShouldCacheCurrentBlock = false;
-        }
-      }
+  protected boolean shouldUpdateStreams(long currentBlockId) {
+    if (mCurrentBlockInStream == null || currentBlockId != mStreamBlockId) {
+      return true;
     }
+    if (mCurrentCacheStream != null) {
+      Preconditions.checkState(mCurrentBlockInStream.remaining() == mCurrentCacheStream.remaining(),
+          "BlockInStream and CacheStream is not sync-ed %d %d", mCurrentBlockInStream.remaining(),
+          mCurrentCacheStream.remaining());
```

Comment:
This is pretty cheap check (should be unnoticeable compared to those IOs). I think it makes me more confident on the code. If you still prefer to removing it, I can do it.